zkevm-circuits/src/evm_circuit/execution/mul_div_mod_shl_shr.rs

```diff
@@ -0,0 +1,351 @@
+    divisor_is_zero: IsZeroGadget<F>,
+    /// Check if remainder is zero for opcode MUL and SHL
+    remainder_is_zero: IsZeroGadget<F>,
+    /// Check if remainder < divisor when divisor != 0
+    remainder_lt_divisor: LtWordGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for MulDivModShlShrGadget<F> {
+    const NAME: &'static str = "MUL_DIV_MOD_SHL_SHR";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::MUL_DIV_MOD_SHL_SHR;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        let is_mul = is_mul(&opcode);
+        let is_div = is_div(&opcode);
+        let is_mod = is_mod(&opcode);
+        let is_shl = is_shl(&opcode);
+        let is_shr = is_shr(&opcode);
+
+        let quotient = cb.query_word();
+        let divisor = cb.query_word();
+        let remainder = cb.query_word();
+        let dividend = cb.query_word();
+        let pop1 = cb.query_word();
+
+        let mul_add_words =
+            MulAddWordsGadget::construct(cb, [&quotient, &divisor, &remainder, &dividend]);
+        let divisor_is_zero = IsZeroGadget::construct(cb, sum::expr(&divisor.cells));
+        let remainder_is_zero = IsZeroGadget::construct(cb, sum::expr(&remainder.cells));
+        let remainder_lt_divisor = LtWordGadget::construct(cb, &remainder, &divisor);
+
+        // Based on different opcode cases, constrain stack pops and pushes as:
+        // - for `MUL`, two pops are quotient and divisor, and push is dividend.
+        // - for `DIV`, two pops are dividend and divisor, and push is quotient.
+        // - for `MOD`, two pops are dividend and divisor, and push is remainder.
+        // - for `SHL`, two pops are shift and quotient, and push is dividend.
+        // - for `SHR`, two pops are shift and dividend, and push is quotient.
+        cb.stack_pop(pop1.expr());
+        cb.stack_pop(
+            (is_mul.clone() + is_div.clone() + is_mod.clone()) * divisor.expr()
+                + is_shl.clone() * quotient.expr()
+                + is_shr.clone() * dividend.expr(),
+        );
+        cb.stack_push(
+            (is_mul.clone() + is_shl.clone()) * dividend.expr()
+                + (is_div.clone() + is_shr.clone())
+                    * quotient.expr()
+                    * (1.expr() - divisor_is_zero.expr())
+                + is_mod.clone() * remainder.expr() * (1.expr() - divisor_is_zero.expr()),
+        );
+
+        cb.require_zero(
+            "remainder < divisor when divisor != 0",
+            (1.expr() - divisor_is_zero.expr()) * (1.expr() - remainder_lt_divisor.expr()),
+        );
+
+        cb.require_zero(
+            "remainder == 0 for both opcode MUL and SHL",
+            (is_mul.clone() + is_shl.clone()) * (1.expr() - remainder_is_zero.expr()),
+        );
+
+        cb.require_zero(
+            "overflow == 0 for opcode DIV, MOD and SHR",
+            (is_div.clone() + is_mod.clone() + is_shr.clone()) * mul_add_words.overflow(),
+        );
+
```

Comment:
I think you're still missing the power of 2 lookup conditions for SHL and SHR?

You can check that pop2 < 256 by checking that the value fits into the first byte with by constraining that RLC(pop2) == pop2.cells[0] btw.

Reply:
It seems that there is no need to lookup Pow2 table. Since I calculated `divisor = 1 << shf0` (if shf0 < 256) in function `assign_exec_step` directly (not sperated to low and high bytes). What do you think?

And I added another contraint to check `pop1 == pop1.cells[0]` for opcode SHL and SHR. Thanks.
```
cb.require_zero(
    "pop1 == pop1.cells[0] when divisor != 0 for opcode SHL and SHR",
    (is_shl.clone() + is_shr.clone())
        * (1.expr() - divisor_is_zero.expr())
        * (pop1.expr() - pop1.cells[0].expr()),
);
```